ApollosProject/apollos-apps
# Adds 'up next' support to GraphQL API

## DESCRIPTION

### What does this PR do, or why is it needed?

This PR adds the ability for ContentSeriesContentItems to query for the "up next" item. That item is the first item that comes after the last item a user completed it. Example: (bold represents items that have been completed

A. B. **C.** D. E. <- `C` would be picked

A. B. C. D. E. <- `A` would be picked

**A. B. C. D. E.** <- `null` would be picked

### What design trade-offs/decisions were made?

There's some code design choices I had to make that will have an impact on performance. Those have been documented through comments in the PR. 

### How do I test this PR?

You can simulate this PR in action by using the query/mutation below.

1. Login to GraphQL via mutations and storing the auth token.
2. Run the "series" query. Note that title and `id` of the upNext item.
3. Plug that `id` into the `readNode` mutation and run it. 
4. Now run the series query again. Note that the result has changed.

```
query series {
  node(id:"ContentSeriesContentItem:b74c2d295a0cf1340232ffa8be856c00"){
    id
    ... on ContentSeriesContentItem {
      upNext {
        id
        title
      }
    }
  }
}

mutation readNode {
  interactWithNode(action:COMPLETE, nodeId:"WeekendContentItem:836a837a1aec93fc163d45315340ff0c"){
    success
  }
}
```

### What automated tests did you write?

## TODO:


## REVIEW:

**Manual QA**


**Code Review: Questions to consider**


> The purpose of PR Review is to _improve the quality of the software._
